v3/client/decrypt_middleware.go

```diff
@@ -11,11 +11,65 @@ import (
 	"github.com/aws/aws-sdk-go-v2/service/s3"
 	"github.com/aws/smithy-go"
 	"github.com/aws/smithy-go/middleware"
-
 	smithyhttp "github.com/aws/smithy-go/transport/http"
+	"mime"
 	"strings"
+	"unicode/utf16"
+	"unicode/utf8"
 )
 
+func customS3Decoder(matDesc string) (decoded string, e error) {
+	// Manually decode S3's non-standard "double encoding"
+	// First, mime decode it:
+	decoder := new(mime.WordDecoder)
+	s, err := decoder.DecodeHeader(matDesc)
+	if err != nil {
+		return "", fmt.Errorf("error while decoding material description: %s\n from S3 object metadata: %w", matDesc, err)
+	}
+	var sb strings.Builder
+
+	skipNext := false
```

Comment:
Is it possible to not use the flag and iterate / increment two bytes ? Not a fan of stateful flags and if statements.

Reply:
Fair point, I tried a few things and couldn't come up with a better solution. `range` doesn't let you write to the index, which would be convenient since we only read two bytes when it's UTF-16. We could always read two bytes at a time, but I don't think that's cleaner since it would need more conditionals to avoid index out of bounds errors.